llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp

```diff
@@ -997,8 +997,17 @@ AArch64LoadStoreOpt::mergePairedInsns(MachineBasicBlock::iterator I,
       //   STRWui %w0, ...
       //   USE %w1
       //   STRWui kill %w1  ; need to clear kill flag when moving STRWui upwards
-      RegOp0.setIsKill(false);
-      RegOp1.setIsKill(false);
+      for (auto It = std::next(I);
+           It != Paired && (RegOp0.isKill() || RegOp1.isKill()); ++It) {
+        auto ClearKill = [](MachineInstr &MI, MachineOperand &MOP,
+                            const TargetRegisterInfo *TRI) {
+          Register Reg = MOP.getReg();
+          if (MI.readsRegister(Reg, TRI) || MI.modifiesRegister(Reg, TRI))
```

Comment:
A `readsRegister` check should be enough I think? I guess you could technically have a "use" by having a subregister def, but for that to make sense we would either need to have another real use within the sequence that will get catched by `MI::readsRegister` anyway or the operands cannot have been a `kill` in the first place.

Reply:
Good point. I pushed a new commit to remove the `modifiesRegister` check.